Newsletter: Implement SelectDropdown for accent color selection

mashikag opened the pull request:
![Screen Shot 2022-10-13 at 18 38 37](https://user-images.githubusercontent.com/2019970/195655124-7591fee9-a1c2-4545-a69b-5ae944a53ae2.png)
![Screen Shot 2022-10-12 at 18 34 37](https://user-images.githubusercontent.com/2019970/195399187-d5329b33-3d5f-413d-a937-747a8890d8bb.png)


#### Proposed Changes

* implement SelectDropdown to allow for selection of predefined accent colors. Combine this with an ability to chose a color via color picker if custom option is selected.

#### Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Go to http://calypso.localhost:3000/setup/newsletterSetup?flow=newsletter

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Closes #68659

mashikag committed:
Newsletter: Implement SelectDropdown for accent color selection

mashikag committed:
Align SelectDropdown's focus outline with the rest of the form. Hide selected option unless the selected option is 'Custom'.

mashikag commented:
With the lack of a response to my above [questions](https://github.com/Automattic/wp-calypso/pull/68996#issuecomment-1276471627) I finished up the PR. I decided to re-use the already existing `SelectDropdown` component. I found it to be the less time consuming approach (in short term). The PR is ready to be reviewed.

mashikag commented on client/components/select-dropdown/style.scss:
```diff
@@ -40,7 +40,7 @@ $compact-header-height: 35;
 
 	.accessible-focus & {
-		border-radius: 2px;
+		border-radius: 4px;
```
I realize that any changes in this file will apply to most `SelectDropdown`s in wp-calypso. I think it just adds the more modern touch to the component (it's 7 years old) and aligns closer with our recent designs. However if one reviewing would prefer I would only apply those for the accent color select-dropdown, I will separate this out.

ebinnion commented:
Apologies for the delay. High level, reusing this select dropdown makes sense to me.

To address your concerns on accessibility, are there ways that we can improve that and improve the component across all of Calypso?

On refreshing the design of the component rather than overriding its styles for our case, I'd prefer the component be updated generally. Then, perhaps let's audit a few different places where the component is used to ensure that things are not out of place. If things do feel out of place, then perhaps we take a pass at generally refreshing the form components.

mashikag commented:
> To address your concerns on accessibility, are there ways that we can improve that and improve the component across all of Calypso?

Good question. I am not an accessibility expert myself. I just did a quick research online and this article seems to make sense: [How to Create an Accessible Custom Select Dropdown in React](https://towardsdev.com/how-to-create-an-accessible-custom-select-dropdown-in-react-9d9858415f10). At the end of the day it seems like I was wrong to say the following:
> The SelectDropdown is based on div element instead of select or input element.

However there are still some accessibility issues that should be addressed, e.g. dropdown options have `aria-role` set to `menuitem` where it should be `option` instead - at least when the parent container is `listbox` like in this case. I will try to address some of them with this PR.

mashikag committed:
Improve accessibility experience for SelectDropdown

mashikag commented:
I updated the PR. I think this is as far as I should go with this one. However I wanted to share my observations in relation to the `SelectDropdown` component.

### My observations in relation to the `SelectDropdown` component:
- Needs to be migrated over to TS
- In `SelectDropdownItem`:
    - It may be more accurate to rename the component to `SelectDropdownOption`
    - There is no need to use anchor (`<a/>`) elements within every dropdown item component
    - Using list item (`<li/>`) element should be sufficient. All `aria-*` tags should be moved over to the list item element. Click handler should also be moved over from the anchor to the list item. If that was to be done one of the a11y rules would for the `li` would be broken:
    > Visible, non-interactive elements with click handlers must have at least one keyboard listener. eslint: [[jsx-a11y/no-noninteractive-element-interactions](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/tree/HEAD/docs/rules/no-noninteractive-element-interactions.md)](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/tree/HEAD/docs/rules/no-noninteractive-element-interactions.md)
    This could be either fixed by disabling the rule and keeping our key-down handler in the `SelectDropdown` component, or slightly modifying the handler and applying it in the `SelectDropdownItem` itself. Personally I would probably prefer to have the 2nd suggestion implemented, as per the eslint rule broken.
- It would be useful to allow for option `value`s to be objects
- In my opinion we should not allow for `path` property on options. It goes against what the `SelectDropdown` is. Instead it introduces menu-like behavior. This also means that `aria-role`s of `listbox` for the dropdown container and `option` for options would be now incorrect.
- I am not sure about the selected item design. Perhaps we should consider not displaying the selected item in the list of options by default, like per our design.
- Perhaps we should unify the styles with those of calypso’s form components? It is not that far off. Other than what I am changing now, i.e.: `border-radius`, `font-weight` for the selected, I noticed the outline on focus also would require to be updated.
- Re-opening the dropdown using keyboard (`SPACE` or `ENTER`) does not trigger a focus on an option until `TAB` or one of arrow keys are pressed.
- Use `button` html native element in place of `div.select-dropdown__header` instead of `role=button` for the `div.select-dropdown__container` (just like in this [[guide](https://towardsdev.com/how-to-create-an-accessible-custom-select-dropdown-in-react-9d9858415f10)](https://towardsdev.com/how-to-create-an-accessible-custom-select-dropdown-in-react-9d9858415f10))

mashikag committed:
Fix unit tests. Revert usage of aria-activedescendant attribute.

mashikag committed:
Use aria-haspopup listbox instead of true as value.

mashikag committed:
Remove empty line between imports

yansern commented:
**Accent Color UI:**
- It is odd to me that Orange & Purple which are the colors we suggest are showing the contrast warning.
- If Orange & Purple was chosen, and I went ahead and pick a dark custom color, the contrast warning doesn't go away.

**SelectDropdown**

It feels very risky to introduce changes to the `SelectDropdown` where if we're going to be careful, we'll have to do an exhaustive test on other places that use this component. Here's a list of places where `SelectDropdown` is used in Calypso:

![image](https://user-images.githubusercontent.com/1287077/196415893-0e361c0a-95fc-43c7-b654-aaf428559c67.png)

Additionally, `SelectDropdown` is also used by `NavTabs` component, and this component is used on these places.

![image](https://user-images.githubusercontent.com/1287077/196421118-614afcbd-c617-4286-aa6b-77b389c17576.png)
 
> - It may be more accurate to rename the component to SelectDropdownOption

I think it is the mixed-use case of `SelectDropdown`, where one is being used more like a navigation menu.

I would argue that:
- for navigation usage, it should be called `NavDropdown`
- for form control usage, it should be called `SelectDropdown`.

And it is also due to this mixed-use case that has led to all of the design/technical considerations have you have listed above.

> - There is no need to use anchor (`<a/>`) elements within every dropdown item component
> - Using list item (`<li/>`) element should be sufficient. All aria-* tags should be moved over to the list item element. Click handler should also be moved over from the anchor to the list item. If that was to be done one of the a11y rules would for the li would be broken:
> - Visible, non-interactive elements with click handlers must have at least one keyboard listener. eslint: [jsx-a11y/no-noninteractive-element-interactions](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/tree/HEAD/docs/rules/no-noninteractive-element-interactions.md)
> 
> - This could be either fixed by disabling the rule and keeping our key-down handler in the SelectDropdown component, or slightly modifying the handler and applying it in the SelectDropdownItem itself. Personally I would probably prefer to have the 2nd suggestion implemented, as per the eslint rule broken.

In short, not using `<a/>` and introduce additional `aria-*` tags to mimic the behaviour of the `<li>` tags.

However, what are the arguments against using `<a/>` though? Would that result in limited freedom in how the HTML within the `item.label` can be presented?

> It would be useful to allow for option values to be objects

While it is possible, it also opens up the risk of failing object comparison tests when trying to determine which one is the current select item.

As the value itself is also used as a `key` in the `DropdownItem`, which either will result in `[Object object]` when stringified, causing keys not to be unique.

Alternatively, it might be safer to introduce optional prop `item.id` or `item.key` so `item.value` can be object, that just comes at the cost of additional conditional checks.

> In my opinion we should not allow for path property on options. It goes against what the SelectDropdown is. Instead it introduces menu-like behavior. This also means that aria-roles of listbox for the dropdown container and option for options would be now incorrect.

I think the use case for this is when Dropdown behaves like menu links, with the ability to say, "Cmd+Click", to open the items in the dropdown in a new tab.

I'd reckon this is used widely in the `NavTab` component.

> I am not sure about the selected item design. Perhaps we should consider not displaying the selected item in the list of options by default, like per our design.

Prior to introducing `showSelectedOption` prop, what was the default behaviour? Are selected option shown or omitted? As long as it behaves the same way as before for other places that uses this component then it is fine.

> Perhaps we should unify the styles with those of calypso’s form components? It is not that far off. Other than what I am changing now, i.e.: border-radius, font-weight for the selected, I noticed the outline on focus also would require to be updated.

On the contrary, it is the stepper framework itself that is overriding the styles of Calypso form components, e.g. [the focus override for text input in stepper stylesheet](https://github.com/Automattic/wp-calypso/blob/trunk/client/landing/stepper/declarative-flow/internals/global.scss#L238).

So overriding the `SelectDropdown` stylesheet within Stepper framework (which is what you're doing right now) is fine.

However, the part where the `border-radius` is changed from `2px` to `4px` globally would cause inconsistency in existing Calypso pages.

> Re-opening the dropdown using keyboard (SPACE or ENTER) does not trigger a focus on an option until TAB or one of arrow keys are pressed.

This might be worth fixing as a separate patch.

----

I'm afraid of approving this PR solely based on the potential side effects of modifying `<SelectDropdown>`.

In short, would it be possible to have a safer implementation of the `<SelectDropdown>` component that doesn't introduce all the risks factor, while allowing room for experimentation?

In my opinion, you could start duplicating this `<SelectDropdown>` component so you have more room for experimentation, and steer this `<SelectDropdown>` component towards the **Form Control** use case.

As for the existing component, it could be renamed to `<NavDropdown>` to better steer this component towards the **Navigation Menu** use case.

As for short-term:
1. Renaming `menuitem` to `option` felt like we're changing one definition to fit another definition when both use cases are valid. Best to leave things as is.
2. Is the removal of Gridicon / MaterialIcon checking safe? Perhaps we should just be checking if the `icon` prop is a React element and let it render. 
3. Changing the border-radius globally would result in inconsistency in Calypso pages. Rather, the border-radius should be an override within Stepper's stylesheet.

mashikag committed:
Try to keep changes in SelectDropdown to minimum. Fix selection of custom color.

mashikag commented:
@yansern wow! What a review! ❤️ ❤️  I did not expect the reviewer to actually go and review my observations too. Perhaps it was not the best place to put them to. 😅 

I agree with many points that you made and I did make an update to the PR to adhere to some of your suggestions. I addressed the following.

### 1. `menuitem` vs `option`

Quoting parts of your above review:
> I think it is the mixed-use case of SelectDropdown, where one is being used more like a navigation menu.
> 
> I would argue that:
> 
> for navigation usage, it should be called NavDropdown
> for form control usage, it should be called SelectDropdown.
> And it is also due to this mixed-use case that has led to all of the design/technical considerations have you have listed above.

and

> > In my opinion we should not allow for path property on options. It goes against what the SelectDropdown is. Instead it introduces menu-like behavior. This also means that aria-roles of listbox for the dropdown container and option for options would be now incorrect.
> 
> I think the use case for this is when Dropdown behaves like menu links, with the ability to say, "Cmd+Click", to open the items in the dropdown in a new tab.
> 
> I'd reckon this is used widely in the NavTab component. 

Great points! 🎯  I reverted my changes where I was changing the `aria-roles` and related. This included reverting changes in unit tests.

### 2. Changing `SelectDropdown` Styles

> > Perhaps we should unify the styles with those of calypso’s form components? It is not that far off. Other than what I am changing now, i.e.: border-radius, font-weight for the selected, I noticed the outline on focus also would require to be updated.
> 
> On the contrary, it is the stepper framework itself that is overriding the styles of Calypso form components, e.g. [the focus override for text input in stepper stylesheet](https://github.com/Automattic/wp-calypso/blob/trunk/client/landing/stepper/declarative-flow/internals/global.scss#L238).
> 
> So overriding the SelectDropdown stylesheet within Stepper framework (which is what you're doing right now) is fine.
> 
> However, the part where the border-radius is changed from 2px to 4px globally would cause inconsistency in existing Calypso pages.

I reverted all changes made to `SelectDropdown` styles and instead I override them for `AccentColorControl` only.

### 3. Custom Color Selection Issue

> If Orange & Purple was chosen, and I went ahead and pick a dark custom color, the contrast warning doesn't go away.

Well spotted. It was a flaw in my implementation. It should work now as expected.

_______ 


‼️ On the other hand I did not address the:

> Is the removal of Gridicon / MaterialIcon checking safe? Perhaps we should just be checking if the icon prop is a React element and let it render.

It is a fundamental change to achieve the goal of this PR. Prior to making the decision to go down this route of reusing the `SelectDropdown` component I did have a private slack discussion with the 2 people who introduced the checks. They see no objections to removing those (I did send you screenshots over in a DM).

mashikag committed:
Fix custom color selection

mashikag committed:
Display custom color icon 4px up from the middle so it appears to be vertically center alligned

mashikag commented:
Hey @yansern ! 👋 

> 1. I have trouble clicking Custom on first click. Once I reopen the dropdown, it works the second time.
> https://user-images.githubusercontent.com/1287077/196868986-9b3461c1-5a8e-4b2c-8204-b241a78fac1a.mp4
> 
> 2. If I pick a custom color, the color retains the name of the preset that was previously chosen. It should just show the hex color.
> https://user-images.githubusercontent.com/1287077/196869373-0bfbf502-8fe6-4b1f-9500-dddcb39e2c52.mp4

Thanks for catching the 1 and 2. This is embarrassing. I removed a `return` statement from [the line](https://github.com/Automattic/wp-calypso/pull/68996/files#diff-7ef7e4486dfb2e270808a116bbcce9231c12e1976dbe82231c43c14c479e5086R82) when debugging another custom color selection issue and never put it back there - the one with contrast message not disappearing on correct custom color selection. 🙈 

> 4. The icon doesn't look vertically aligned center.
> image

I also addressed the issue no. 4. It should look better now.
____
!! I did not address issue no. 3

> For colors coming from our own preset, e.g. orange & purple, I would suggest not showing the contrast checker warning. That would be weird as to why we're recommending colors that breaks contrast itself.

I do not think this is the right way to go about this one. This is just masking the real issue. In my opinion either the pre-defined colors need to be re-evaluated to include ones with an acceptable color contrast ratio, or in case they are incorrectly marked as not meeting the contrast ratio requirements then we should re-evaluate and fix our algorithm. @saygunnyc what do you think about this one? Based on our algorithm the **Luminous vivid orange** and **Vivid purple** do not meet the color contrast ratio requirements.

saygunnyc commented:
> I do not think this is the right way to go about this one. This is just masking the real issue. 
💯 

I originally intended us to show the 2 most popular colors that live on the default palette.
Which ones are popular and would be accessible?

<img width="262" alt="Screen Shot 2022-10-20 at 12 31 27 PM" src="https://user-images.githubusercontent.com/78876323/196912283-160f9ff1-715f-455c-8868-ff3d41408146.png">

mashikag commented:
> Which ones are popular and would be accessible?

Vivid red - #CF2E2E
Vivid purple - #9B51E0

At the end seems like **Vivid purple** is accessible, it is just the figma design contained a different shade of that color (#925CD9) which resolves to inaccessible. This means that only **Luminous vivid orange** needs to be replaced and the only option seems to be **Vivid red**. Does that sound ok @saygunnyc ?

saygunnyc commented:
Good catch @mashikag! Yes, let's do that for now. 
Just noting that we might end up setting custom values at a later point.

mashikag committed:
Replace color options not meeting contrast ratio requirements with ones meeting those

yansern reviewed (approved):
The color picker works as expected. Approving PR.
